### ISSUE TITLE
Weighting in scalar and dmc data files differ

### ISSUE THREAD

**Issue by jtkrogel** [QMCPACK/qmcpack#1443]
It has been noticed more than once that averages based on scalar.dat and dmc.dat files differ.  This is due to different weighting schemes used during accumulation of the data.

Let W(s,p) and E(s,p) be the weight and local energy of the p'th walker at MC step s.  The total population at step s is P(s), i.e. p in [1,P(s)] and the number of steps is S.  A block is a set of B consecutive steps, indexed by b in [1,S/B].

The energy values written to dmc.dat are per step according to the formula:
E_dmc(s) = (sum_p<P(s) W(s,p)*E(s,p)) / (sum_p<P(s) W(s,p))

The energy values written to scalar.dat are per block according to the formula:
E_scalar(b) = (sum_s<B sum_p<P(b+s) W(b+s,p)*E(b+s,p)) / (sum_s<B sum_p<P(b+s) W(b+s,p))

The dmc.dat way is identical to what is described in Luchow 2011 Eq. 26 (https://doi.org/10.1002/wcms.40), while the scalar.dat way is similar to what is described in Umrigar 1993 Eq. 19 (https://doi.org/10.1063/1.465195) with T being the number of steps in a block (B) rather than the entire run (which offers yet a third way to get the mean).

So we have three ways to obtain the mean with a finite fluctuating population, that differ primarily in the window of time used to represent a realization of f(R)=Psi_T*Psi_0, with dmc.dat being instantaneous (single time slice), scalar.dat being an average over a short time window, and the method outlined in Umrigar 1993 being an average over the entire equilibrated run (please correct if I have missed something). 

Should we be concerned about the difference or just document it?  

If I am right, these forms all agree in the large population limit, but contain different statistical biases for finite populations.  It may be obvious to someone else which has the smallest statistical bias...

**Comment by jtkrogel** [comment]
I had a discussion with @lshulen and @kgasperich over a year ago on this, so they may have some thoughts.

**Comment by ye-luo** [comment]
It is wired to me that in the code, the concept of total population and total population weight is mixed.
The code sometimes use the number of walkers after branching as the total population weight.
In principle, the rounding effect should cancel but I feel unsafe using it.
Instead, I'm thinking that all the statistics should be accumulated before the branching and taking account each walker's weight.

My understanding is
Total population weight PW(s) = sum_p<W(s,p)>
E_dmc(s) = (sum_p<W(s,p)*E(s,p)>) / (sum_p<W(s,p)>)
Then
E_scalar(b) = (sum_s<sum_p<W(b+s,p)*E(b+s,p)>>) / (sum_s<sum_p<W(b+s,p)>>)
                   = (sum_s<P(b+s)*E_dmc(b+s)>)/(sum_s<PW(b+s)>)
In this way, E_dmc and E_scalar is consistent.
Then the post processing tools needs to handle all sorts of estimator correction.

**Comment by lshulen** [comment]
My general preference is along the lines of what @ye-luo wrote.  I'll go back and review your references to see it it changes my mind.

**Comment by jtkrogel** [comment]
What @ye-luo writes is essentially Umrigar 1993 eq. 19.

Also see Foulkes 2001 eq. 3.53 (https://doi.org/10.1103/RevModPhys.73.33), the line before transforming the integral into a sum.  In the absence of the limit, this defines a DMC energy at time t, or equivalently, at step s:

E_DMC(s) = int_R f(R,s)E(R) / int_R f(R,s)

In the limit of large population, this integral is exactly represented at a single time t/step s.  Once f is stationary, that is f(R,s+1)=f(R,s), then each E_DMC(s) is identical, or statistically can be viewed as an independent sample of the same quantity, and we get:

<E_DMC> = 1/S sum_s<S E_DMC(s)

This is what is described in Luchow 2011 Eq. 26 and this is what we do for dmc.dat.

If instead one takes a time average of f as being equivalent to f in the long time limit, i.e.

f(R,inf) = 1/T int_t f(R,t) , integral over t in [t_equil, t_equil+T]

then with T being the time over the whole run and substituting into the first formula I wrote, the DMC energy is

E_DMC = int_s int_R f(R,s)E(R) / int_s int_R f(R,s)

Evaluating this formula literally gives what Ye wrote for the DMC energy and also Umrigar 1993 eq 26.  [As an aside, one thing I haven't worked out for myself yet, is how the above formula deals with changes in the norm of f with time (the difference in norm is the difference in the sum of the weights at different times)]

In other words, these approaches only differ in how the statistical representation of f is viewed (single point in time, average over the time window of a block, average over the time in the entire run) and they agree in the large population limit.

I'm not really advocating for any of them, just pointing out that they are equivalent in the large population limit.

Unfortunately, I'm not sure how to rigorously select one over another.  The difference between them should be a statistical bias that depends on population size, and the best approach is the one with the smallest bias.  However, when attempting evaluate the difference empirically we will not be able to measure this bias in isolation of the population control bias.

**Comment by zenandrea** [comment]
Hi guys,
my opinion is that for each value of s, the E_dmc(s) defined by @jtkrogel in the first message
should have a weight: 
W_step(s) = sum_p<P(s) W(s,p)
and the qmca tool should compute the weighted average, that is:
E_dmc = {\sum_s W_step(s) E_dmc(s) \over \sum_s W_step(s) }.

When the blocks (scalar.dat file) are instead considered, the weight associated to a block energy E_scalar(b) should be: 
W_block(b) = sum_s<B sum_p<P(b+s) W(b+s,p).
Then, the qmca tool should evaluate the energy again via a weighted average:
E_dmc = {\sum_b W_block(b) E_scalar(b) \over \sum_b W_block(b) }.

In this way the estimations from dmc.dat and scalar.dat should be identical.

Moreover, I think that neither of the expressions discussed so far in this topic are actually similar to the Umrigar's estimator.
The key point in Umrigar's estimator is that it uses Tp correcting factors, namely there is the multiplication of the weights by the factor \Pi(s,Tp), see eq. 14 in the paper. As they explains, a large enough Tp projects out the population control bias, but if Tp is too large the cost is a larger variance.
Ideally, one would like to plot the estimated dmc energy E_dmc(Tp) as a number of correcting factors Tp, observing that E_dmc(Tp) first decrease and then becomes flat, which is the population unbiassed value of E_dmc.

**Comment by ye-luo** [comment]
@jtkrogel Actually I didn't understand the notion of `sum_p<P(s) W(s,p)` in your original note. Does it mean sum over p in P(s) instead of P(s) W(s,p) multiplied together?

**Comment by jtkrogel** [comment]
@ye-luo It means \sum_{p=1}^{P(s)} W(s,p).  GitHub would be much better w/ LaTeX support.

@zenandrea I should have been more clear about the population control correction factor.  In the above, I meant \Pi(s,Tp)=1, or Tp=0, which is a choice for Umrigar's mixed estimator without population control correction.

**Comment by ye-luo** [comment]
Measuring quantities with weight before branching and measuring quantities without weights after branching are numerically different due to random numbers and rounding. They should be statistically equivalent.
dmc.dat does the former, scalar.dat does the latter. This explains dmc.dat weight can be float numbers but scalar.dat is always integer.

I have changed the behavior in batched driver. all the statistics are measured before branching with weight.

```
yeluo@epyc-server:~/opt/qmcpack/build_clang_real_debug/tests/solids/diamondC_2x1x1_pp/deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1$ cat det_qmc_short_vmc_dmc.s001.dmc.dat
# Index          LocalEnergy            Variance              Weight        NumOfWalkers      AvgSentWalkers         TrialEnergy             DiffEff      LivingFraction
         0   -1.9215165346e+01    5.6843418861e-14    1.0002887511e+00    1.0000000000e+00    0.0000000000e+00   -1.9099681568e+01    9.1371423069e-01    1.0000000000e+00
         1   -1.9395946598e+01    0.0000000000e+00    1.0007019117e+00    1.0000000000e+00    0.0000000000e+00   -1.9215165346e+01    7.8005143109e-01    1.0000000000e+00
         2   -2.0113079062e+01   -5.6843418861e-14    1.0014853924e+00    1.0000000000e+00    0.0000000000e+00   -1.9395946598e+01    1.0000000000e+00    1.0000000000e+00
         3   -1.9770961380e+01    0.0000000000e+00    9.9924805124e-01    1.0000000000e+00    0.0000000000e+00   -2.0113079062e+01    1.0000000000e+00    1.0000000000e+00
         4   -1.9892444946e+01    5.6843418861e-14    1.0002736225e+00    1.0000000000e+00    0.0000000000e+00   -1.9770961380e+01    1.0000000000e+00    1.0000000000e+00
         5   -1.8555021304e+01    1.1368683772e-13    9.9694793159e-01    1.0000000000e+00    0.0000000000e+00   -1.9892444946e+01    1.0000000000e+00    1.0000000000e+00
         6   -1.9376806061e+01   -5.6843418861e-14    1.0019034341e+00    1.0000000000e+00    0.0000000000e+00   -1.8555021304e+01    1.0000000000e+00    1.0000000000e+00
         7   -1.9195524388e+01    1.1368683772e-13    9.9957561952e-01    1.0000000000e+00    0.0000000000e+00   -1.9376806061e+01    1.0000000000e+00    1.0000000000e+00
         8   -1.9084864687e+01    5.6843418861e-14    9.9973959187e-01    1.0000000000e+00    0.0000000000e+00   -1.9195524388e+01    1.0000000000e+00    1.0000000000e+00

yeluo@epyc-server:~/opt/qmcpack/build_clang_real_debug/tests/solids/diamondC_2x1x1_pp/deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1$ cat det_qmc_short_vmc_dmc.s001.scalar.dat
#   index    LocalEnergy         LocalEnergy_sq      LocalPotential      Kinetic             ElecElec            IonIon              LocalECP            NonLocalECP         BlockWeight         BlockCPU            AcceptRatio         
         0   -1.9574920294e+01    3.8332790041e+02   -5.4157440357e+01    3.4582520063e+01   -6.0322801613e+00   -2.5551326884e+01   -1.2390951280e+01   -1.0182882032e+01    3.0024760552e+00    1.7537117004e-02    9.5833333333e-01
         1   -1.9406962314e+01    3.7699452898e+02   -4.9499952347e+01    3.0092990033e+01   -5.8763337744e+00   -2.5551326884e+01   -1.1939788771e+01   -6.1325029179e+00    2.9964696054e+00    1.7536163330e-02    1.0000000000e+00
         2   -1.9219180061e+01    3.6939137249e+02   -4.8799579468e+01    2.9580399407e+01   -5.5403915017e+00   -2.5551326884e+01   -1.3169357437e+01   -4.5385036449e+00    3.0012186455e+00    1.7354011536e-02    1.0000000000e+00
```
There is no more difference between scalar.dat and dmc.dat
